File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20141230_Dash6/Solution.java
import java.util.Scanner;

public class Solution {
  public static int[] partition(int[] ar) {
    int[] res = new int[ar.length];
    int idx = 0;
    for ( int i = 0; i < ar.length; ++i ) {
      if ( ar[i] < ar[0] )  {
        res[idx++] = ar[i];
      }
    }
    res[idx++] = ar[0];
    for ( int i = 0; i < ar.length; ++i ) {
      if ( ar[0] < ar[i] )  {
        res[idx++] = ar[i];
      }
    }
    return res;
  }

  public static void main(final String[] args)  {
    Scanner sc = new Scanner(System.in);
    int n = Integer.parseInt(sc.nextLine());
    String[] inps = sc.nextLine().split(" ");
    int[] ar = new int[inps.length];
    for ( int i = 0; i < ar.length; ++i ) {
      ar[i] = Integer.parseInt(inps[i]);
    }
    int[] res = partition(ar);
    System.out.print(res[0]);
    for ( int i = 1; i < res.length; ++i )
      System.out.print(" " + res[i]);
    System.out.println();
  }
}

File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20141230_Dash6/solution.c
#include <stdio.h>
#include <stdlib.h>

void partition(int* ar, int n, int** res)  {
  int pivot = *ar;
  int i = 0, idx = 0;
  for ( i = 0; i < n; ++i ) {
    if ( *(ar + i) < pivot )  {
      *(*res + idx++) = *(ar + i);
    }
  }
  *(*res + idx++) = pivot;
  for ( i = 0; i < n; ++i ) {
    if ( pivot < *(ar + i) )  {
      *(*res + idx++) = *(ar + i);
    }
  }
}

int main()  {
  int n = 0;
  scanf("%d", &n);
  int* ar = (int*) malloc(sizeof(int) * n);
  int* res = (int*) malloc(sizeof(int) * n);
  int i = 0;
  for ( i = 0; i < n; ++i ) {
    scanf("%d", ar + i);
  }
  partition(ar, n, &res);

  printf("%d", *res);
  for ( i = 1; i < n; ++i )
    printf(" %d", *(res + i));
  printf("\n");

  free(ar);
  free(res);
  return 0;
}
